core/drivers/imx_ele.c

```diff
@@ -0,0 +1,576 @@
+	else
+		pa = (paddr_t)buffer;
+
+	reg_pair_from_64((uint64_t)pa, &cmd.addr_msb, &cmd.addr_lsb);
+	cmd.size = size;
+
+	/*
+	 * Check the current TRNG state of the ELE. The TRNG must be
+	 * started with a command earlier in the boot to allow the TRNG
+	 * to generate enough entropy.
+	 */
+	while (imx_ele_rng_get_trng_state() == TEE_ERROR_BUSY)
+		if (timeout_elapsed(timeout))
+			return TEE_ERROR_BAD_STATE;
+
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	update_crc(&msg);
+
+	if (cpu_mmu_enabled())
+		cache_operation(TEE_CACHEFLUSH, (void *)buffer, size);
+
+	res = imx_ele_call(&msg);
+
+	if (cpu_mmu_enabled())
+		cache_operation(TEE_CACHEINVALIDATE, (void *)buffer, size);
+
+	return res;
+}
+
+unsigned long plat_get_aslr_seed(void)
+{
+	unsigned long aslr __aligned_ele = 0;
+
+	if (imx_ele_rng_get_random((uint8_t *)&aslr, sizeof(aslr)))
+		panic("Fail to the seed the ASLR");
+
+	return aslr;
+}
+
+#define UID_SIZE (4 * sizeof(uint32_t))
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
+	uint32_t session_handle = 0;
+	/*
+	 * The die ID must be cached because some board configuration prevents
+	 * the MU to be used by OPTEE at runtime.
+	 */
+	static struct session_get_device_info_rsp rsp;
+
+	if (rsp.rsp_code)
+		goto out;
+
+	if (imx_ele_session_open(&session_handle))
+		goto err;
+
+	if (imx_ele_session_get_device_info(session_handle, &rsp))
+		goto err;
+
+	if (imx_ele_session_close(session_handle))
+		goto err;
+
+out:
+	/*
+	 * In the device info array return by the ELE, the words 2, 3, 4 and 5
+	 * are the device UID.
+	 */
+	memcpy(buffer, rsp.chip_uid, MIN(UID_SIZE, len));
+
+	return 0;
+err:
+	panic("Fail to get the device UID");
+}
+
+#if defined(CFG_MX93)
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	static uint8_t key[HW_UNIQUE_KEY_LENGTH] __aligned_ele;
+	static bool is_fetched;
+	uint32_t msb = 0;
+	uint32_t lsb = 0;
+	struct key_derive_cmd {
+		uint32_t key_addr_msb;
+		uint32_t key_addr_lsb;
+		uint16_t key_size;
+		uint8_t res;
+	} __packed cmd = { };
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_BASELINE,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_DERIVE_KEY,
+	};
+
+	if (is_fetched)
+		goto out;
+
+	/*
+	 * Intermediate msb and lsb values are needed. Directly using
+	 * key_addr_msb and key_addr_lsb might be unaligned because of the
+	 * __packed attribute of key_derive_cmd {}
+	 */
+	reg_pair_from_64((uint64_t)virt_to_phys(key), &msb, &lsb);
+
+	cmd.key_addr_lsb = lsb;
+	cmd.key_addr_msb = msb;
+	cmd.key_size = sizeof(key);
+
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	cache_operation(TEE_CACHEFLUSH, key, HW_UNIQUE_KEY_LENGTH);
+
+	res = imx_ele_call(&msg);
+	if (res) {
+		EMSG("Failed to get the HUK");
+		panic();
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, key, HW_UNIQUE_KEY_LENGTH);
+out:
+	memcpy(hwkey->data, key, HW_UNIQUE_KEY_LENGTH);
+	is_fetched = true;
```

Comment:
nitpicking: move to before `out` branch label

Reply:
Ok